sstables/sstables.cc

```diff
@@ -2138,12 +2142,27 @@ sstable::make_reader_v1(
         streamed_mutation::forwarding fwd,
         mutation_reader::forwarding fwd_mr,
         read_monitor& mon) {
-    schema = maybe_reverse(std::move(schema), slice);
+    auto max_result_size = permit.max_result_size();
+    if (_version >= version_types::mc) {
+        auto reader = downgrade_to_v1(mx::make_reader(shared_from_this(), schema, std::move(permit), range, slice, pc, std::move(trace_state), fwd, fwd_mr, mon));
+        if (*reader.schema() != *schema) {
+            // The reader has modified the provided schema.
+            // Currently this can happen in only one scenario: we perform a reversed query,
+            // thus provided the reader a reversed slice and a reversed schema, but the reader rejected
+            // the reversing request and tells us to perform the reversing ourselves (the data from the reader
+            // will return in forward order).
```

Comment:
Under what conditions can the reader (silently) *reject* to do what it is asked to do?

Reply:
It's not silent, it's explicitly specified in the interface of `mx::make_reader`:
```
// Precondition: if the slice is reversed, the schema must be reversed as well.
// Reversed slices must be provided in the 'half-reversed' format (the order of ranges
// being reversed, but the ranges themselves are not).
//
// If the slice is reversed then:
// - if this is a single-partition read (range.is_singular()), each partition
//   in the returned fragment stream will be reversed - i.e. ordered according to
//   the reversed schema. The reader's schema will be the provided schema.
//   In this mode fast-forwarding is not supported (FIXME);
// - otherwise, the data will be returned in non-reversed order and the reader's
//   schema will be non-reversed - i.e. it will be the reverse of the provided schema
//   (since the provided schema is already reversed according to the precondition).
//   In this case the caller is responsible for reversing the fragment stream
//   themselves.
flat_mutation_reader_v2 make_reader(
```
(see sstables/mx/reader.hh) which also answers your question.

Reply:
I think this is suprising and we should do this here instead create a reversing reader and pass non-reversed slice if the case.

Reply:
That's what I wanted to do, but found that I can't (at least I can't do it easily) because there is nowhere to store the slice, and I didn't want to stuff another intermediate reader in between (there are too many of those already for my taste). If I do this directly in the mx reader, it's easy to store the slice as you can see: in the reader itself.

I think it shouldn't cause any surprises if one reads documentation comments which I diligently wrote. And hopefully we'll be able to remove these workarounds soon, allowing us to simplify the interface.

Reply:
Your suggestion in the other comment gave me an idea: provide two `make_reader` functions, one with `const partition_slice&` param, one with `partition_slice&&` which would store the slice. Both would require that if the range is not singular, the slice is not reversed (and we would have an assertion inside), and it's kinda sad that we would have two functions, but maybe it's better than the current complex interface alternative. What do you think?

Reply:
Yes, this is what I wanted to suggest too.